### ISSUE TITLE
Failed to initialize scheme

### ISSUE THREAD

**Issue by timcan** [xvik/guice-persist-orient#7]
Hi,
I'm trying to configure and bootstrap  framework as part of a TestNG test:
```
@Guice(modules = {PersistenceModule.class})
public class PlayerRepositoryTests {

    @Inject
    PersistService persistService;

    @Inject
    PlayerRepository playerRepository;

    @BeforeTest
    public void persistenceService_isStarted() {
        persistService.start();
    }

    @BeforeTest
    public void repoIsInjected() {
        Assert.assertNotNull(playerRepository);
    }

    @Test
    public void testFindAll() {
        List<Player> result = playerRepository.findAll();
        Assert.assertNotNull(result);
        Assert.assertEquals(0, result.size());
    }
}
```

Per the guide provided on GitHub, I install the appropriate modules with:
```
public class PersistenceModule extends AbstractModule {

    private static final Logger LOG = LoggerFactory.getLogger( PersistenceModule.class );

    @Named("database.url")
    private String url = "memory:acmedb";

    @Named("database.user")
    private String user = "admin";

    @Named("database.password")
    private String password = "admin";

    @Override
    protected void configure() {

        OrientModule orientModule = new OrientModule(url,user,password);
        install(orientModule);

        AutoScanSchemeModule schemeModule = new AutoScanSchemeModule("com.acme.domain.entity");
        install(schemeModule);

        RepositoryModule repositoryModule = new RepositoryModule();
        repositoryModule.defaultConnectionType(DbType.GRAPH);
        install(repositoryModule);
    }
}
```

Define a single VertexType entity in the package to scan:
```
@Persistent
@VertexType
public class Player {

    private String firstName;
    private String lastName;

    @Id
    private String id;
    @Version
    private Long version;
    // 
}
```

And a simple repository definition:
```
@Transactional
@ProvidedBy(DynamicSingletonProvider.class)
public interface PlayerRepository {

    @Query("select from Player")
    public List<Player> findAll();
}
```

I'm getting the following the following stacktrace when I run the test:
```
Jun 16, 2015 9:58:03 PM com.orientechnologies.common.log.OLogManager log
INFO: OrientDB auto-config DISKCACHE=10,695MB (heap=3,641MB os=16,384MB disk=135,208MB)
ru.vyarus.guice.persist.orient.db.scheme.SchemeInitializationException: Failed to initialize scheme
	at ru.vyarus.guice.persist.orient.db.DatabaseManager.start(DatabaseManager.java:99)
	at com.nextlevel.repo.PlayerRepositoryTests.persistenceService_isStarted(PlayerRepositoryTests.java:43)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:84)
	at org.testng.internal.Invoker.invokeConfigurationMethod(Invoker.java:564)
	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:213)
	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:138)
	at org.testng.TestRunner.beforeRun(TestRunner.java:641)
	at org.testng.TestRunner.run(TestRunner.java:609)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:348)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:343)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:305)
	at org.testng.SuiteRunner.run(SuiteRunner.java:254)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1224)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1149)
	at org.testng.TestNG.run(TestNG.java:1057)
	at org.testng.remote.RemoteTestNG.run(RemoteTestNG.java:111)
	at org.testng.remote.RemoteTestNG.initAndRun(RemoteTestNG.java:204)
	at org.testng.remote.RemoteTestNG.main(RemoteTestNG.java:175)
	at org.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:125)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:134)
Caused by: ru.vyarus.guice.persist.orient.db.scheme.SchemeInitializationException: Failed to register model class com.nextlevel.domain.entity.Player
	at ru.vyarus.guice.persist.orient.db.scheme.initializer.ObjectSchemeInitializer.register(ObjectSchemeInitializer.java:74)
	at ru.vyarus.guice.persist.orient.db.scheme.impl.AbstractObjectInitializer.registerClasses(AbstractObjectInitializer.java:88)
	at ru.vyarus.guice.persist.orient.db.scheme.impl.AbstractObjectInitializer.initialize(AbstractObjectInitializer.java:52)
	at ru.vyarus.guice.persist.orient.db.DatabaseManager$2.execute(DatabaseManager.java:94)
	at ru.vyarus.guice.persist.orient.db.DatabaseManager$2.execute(DatabaseManager.java:91)
	at ru.vyarus.guice.persist.orient.db.transaction.template.TxTemplate.doInTransaction(TxTemplate.java:58)
	at ru.vyarus.guice.persist.orient.db.DatabaseManager.start(DatabaseManager.java:91)
	... 29 more
Caused by: ru.vyarus.guice.persist.orient.db.scheme.SchemeInitializationException: Failed to create extension VertexTypeExtension declared by annotation @VertexType on class Player
	at ru.vyarus.guice.persist.orient.db.scheme.initializer.core.ext.ExtensionsDescriptorFactory.createTypeExtension(ExtensionsDescriptorFactory.java:72)
	at ru.vyarus.guice.persist.orient.db.scheme.initializer.core.ext.ExtensionsDescriptorFactory.prepareTypeExtensions(ExtensionsDescriptorFactory.java:60)
	at ru.vyarus.guice.persist.orient.db.scheme.initializer.core.ext.ExtensionsDescriptorFactory.resolveExtensions(ExtensionsDescriptorFactory.java:43)
	at ru.vyarus.guice.persist.orient.db.scheme.initializer.ObjectSchemeInitializer.processType(ObjectSchemeInitializer.java:90)
	at ru.vyarus.guice.persist.orient.db.scheme.initializer.ObjectSchemeInitializer.register(ObjectSchemeInitializer.java:71)
	... 35 more
Caused by: com.google.inject.ConfigurationException: Guice configuration errors:

1) Unable to create binding for ru.vyarus.guice.persist.orient.db.DatabaseManager. It was already configured on one or more child injectors or private modules
    (bound by a just-in-time binding)
  If it was in a PrivateModule, did you forget to expose the binding?
  while locating ru.vyarus.guice.persist.orient.db.DatabaseManager
    for parameter 0 at ru.vyarus.guice.persist.orient.db.scheme.initializer.ext.type.vertex.VertexTypeExtension.<init>(VertexTypeExtension.java:29)
  while locating ru.vyarus.guice.persist.orient.db.scheme.initializer.ext.type.vertex.VertexTypeExtension

1 error
	at com.google.inject.internal.InjectorImpl.getProvider(InjectorImpl.java:1042)
	at com.google.inject.internal.InjectorImpl.getProvider(InjectorImpl.java:1001)
	at com.google.inject.internal.InjectorImpl.getInstance(InjectorImpl.java:1051)
	at ru.vyarus.guice.persist.orient.db.scheme.initializer.core.ext.ExtensionsDescriptorFactory.createTypeExtension(ExtensionsDescriptorFactory.java:70)
	... 39 more
```

What am I missing or doing wrong? Appreciate any insight or help you can provide. Thanks.

**Comment by xvik** [comment]
Hi Tim,
Your configuration looks good. 
The error is strange: it looks like guice was trying to create DatabaseManager two times.
I'm using spock (which is junit based) for tests and never have such problem. 

Could you please send me this example project so I could reproduce the problem?
Even if you can't, I will try to setup testng test in the next few days.

**Comment by timcan** [comment]
Hi Vy,

No problem. I will extract to a project snippet and forward it to you. Probably a good idea to see if the test works in a bare project too. Thanks for the quick response, most appreciated.

**Comment by timcan** [comment]
I should have noted more specifically that the stacktrace was caused by running the test in IntelliJ IDEA 14.0.3. When I run from the command line, I get a different stacktrace:
```
Running com.acme.repo.PlayerRepositoryTests
Configuring TestNG with: TestNG652Configurator
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.926 sec <<< FAILURE! - in com.acme.repo.PlayerRepositoryTests
savePlayer(com.acme.repo.PlayerRepositoryTests)  Time elapsed: 0.012 sec  <<< FAILURE!
java.lang.AbstractMethodError: com.acme.repo.PlayerRepository$GuiceDynamicClass.save(Ljava/lang/String;)Lcom/acme/entity/Player;
        at com.acme.repo.PlayerRepositoryTests.savePlayer(PlayerRepositoryTests.java:42)

Results :

Failed tests: 
  PlayerRepositoryTests.savePlayer:42 AbstractMethod com.acme.repo.PlayerReposit...

Tests run: 1, Failures: 1, Errors: 0, Skipped: 0
```

**Comment by xvik** [comment]
It appears to be a very serious problem.

TestNG guice support implementation use parent injector, so your module will always be created as child injector.
First problem (with scheme initializer) was caused by Injector instance declarations: when type is not declared in module (so will be created with JIT) and it has dependency on Injector, it will always receive root injector. If type is declared in a module and requires Injector then child injector will be set.
I fixed such places and it would be solved in the next release.

Another problem is caused by guice design (your last stack trace): even if type is declared in child injector's module, but it has no dependencies on other types in child injector, binding would be created in parent injector ([see](https://github.com/google/guice/issues/893)). 
This makes repositories useless: repositories are using guice aop, so they don't have explicit (guice) dependencies. As a result, guice will create bindings for repositories in parent injector, where aop interceptors are not registered. On execution you will see an abstract method call error (like in your last comment).

I can't do anything with the last problem. The only solution in such case is to disable JIT, but currently library rely on JIT a lot and can't work without JIT.
 
I can only suggest a few options:
1. Create injector manually instead of using (@Guice) annotation (maybe there are alternative extensions available)
1. Use spock. I'm always using spock because it's much less verbose then pure java solutions. All  tests in the library were written with spock, so you will have a lot of examples.

If you have any ideas please tell me

**Comment by timcan** [comment]
Taking your advice, I went with manually creating the injector and that appears to work fine. Per the Helper.groovy example, I changed the PlayerRepository to extend ObjectCrud<T>. Now I'm getting a different stacktrace when I attempt to change an attribute on the entity after it has been persisted:
```
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Jun 24, 2015 10:39:10 PM com.orientechnologies.common.log.OLogManager log
INFO: OrientDB auto-config DISKCACHE=10,695MB (heap=3,641MB os=16,384MB disk=135,099MB)
com.orientechnologies.orient.core.exception.ODatabaseException: Database instance is not set in current thread. Assure to set it with: ODatabaseRecordThreadLocal.INSTANCE.set(db);
	at com.orientechnologies.orient.core.db.ODatabaseRecordThreadLocal.get(ODatabaseRecordThreadLocal.java:54)
	at com.orientechnologies.orient.object.enhancement.OObjectProxyMethodHandler.getDatabase(OObjectProxyMethodHandler.java:852)
	at com.orientechnologies.orient.object.enhancement.OObjectProxyMethodHandler.setValue(OObjectProxyMethodHandler.java:807)
	at com.orientechnologies.orient.object.enhancement.OObjectProxyMethodHandler.manageSetMethod(OObjectProxyMethodHandler.java:731)
	at com.orientechnologies.orient.object.enhancement.OObjectProxyMethodHandler.invoke(OObjectProxyMethodHandler.java:113)
	at com.acme.entity.Player_$$_javassist_0.setLastName(Player_$$_javassist_0.java)
	at com.acme.repo.PlayerRepositoryTests.savePlayer(PlayerRepositoryTests.java:37)
```

Here's the injector code:
```
    @BeforeClass(alwaysRun = true)
    public void configureBindings() {
        Injector injector = com.google.inject.Guice.createInjector(new PersistenceModule());
        persistService = injector.getInstance(PersistService.class);
        playerRepository = injector.getInstance(PlayerRepository.class);
        persistService.start();
    }
```

And the modified the test method:
```
    @Test(groups={"unit"})
    public void savePlayer() {

        Player p = playerRepository.save("Sheridan");
        Assert.assertNotNull(p);
        Assert.assertEquals(p.getLastName(),"Sheridan");

        List<Player> result = playerRepository.finalAll();
        Assert.assertNotNull(result);
        Assert.assertEquals(result.size(),1);

        p.setLastName("Burk");
        p = playerRepository.save(p);
        Assert.assertEquals(p.getLastName(),"Burk");
    }
```

**Comment by xvik** [comment]
It;s a common issue: you should always work with orient objects under transaction.
When you query for an object it returns you not pure pojo but proxy object, which require connection for  some operations. I can tell that object proxies behaviour outside of transaction changes few times even between 2.0 and 2.0.10 so you shouldn't assume behaviour outside of transaction.
Bu the way the same rule for documents and vertexes/edges.. always work with them under transaction.

In your case transaction scope is limited to repository method. Generally you should simply annotate service (or other upper abstraction) with @Transactional to enlarge scope.

In unit test you can't use annotation, so you should either create test bean and encapsulate logic there (and set transactional on this bean)  or use PersistentContext<connection type>.doWithTransaction inside test to explicitly define transaction scope,

When you need to pass object outside, for example, you want to serialize it to json with jax-rs, you should detach object (under transaction). You can do it using ObjectCrud.detach(object) - it will return pure pojo.

There is special annotation @DetachResults, which you can use on repository methods to automatically detach query results (and get, for example, list of pojos instead of list of proxies).

You can then modify pojo and save it  - it will work fine (but your pojo must have @Id property so orient could recognize object id).

And the last moment, as I remember you use @VertexType on pojo, so you want to work with graph api. In new version there will be ObjectVertexCrud mixin which should be used instead of ObjectCrud for graph objects. The only problem with ObjectCrud for vertex types are delete methods: it use object api to delete, which does not check graph consistency and so you edges to node will remain after remove. ObjectVertexCrud fixes that. 
New version should be release within a week. Or you can build snapshot locally (checkout and run gradlew install).

Regarding your initial issue: I was a bit wrong, it is not impossible to solve it, but I'm still looking for a good solution.

**Comment by timcan** [comment]
Thank you so much for all that information - very helpful and I think things are getting better :+1:  .. so with your suggestions, using the latest snapshot, my repository now extends ObjectVertextCrud (nice!):
```
@ProvidedBy(DynamicSingletonProvider.class)
public abstract class PlayerRepository implements ObjectVertexCrud<Player> {
}
```
Created an transactional persistence bean that has the repository injected into it:
```
@Transactional
public class ParticipantPersister {

    @Inject
    private PlayerRepository playerRepository;

    public Player getPlayer(final String playerId) {
        final Player found = playerRepository.get(playerId);
        return playerRepository.detach(found);
    }

    public Player savePlayer(final Player player) {
        final Player saved = playerRepository.save(player);
        return playerRepository.detach(saved);
    }

    public void deletePlayer(final String playerId) {
        Preconditions.checkArgument(!Strings.isNullOrEmpty(playerId),"playerId is required");
        playerRepository.delete(playerId);
    }

    public void deletePlayer(final Player player) {
        Preconditions.checkNotNull(player,"player is required");
        playerRepository.delete(player);
    }
}
```
And modified the unit tests accordingly:
```
public class ParticipantPersisterTests {

    PersistService persistService;
    ParticipantPersister participantPersister;

    @BeforeClass(alwaysRun = true)
    public void configureBindings() {
        Injector injector = Guice.createInjector(new PersistenceModule());
        persistService = injector.getInstance(PersistService.class);
        participantPersister = injector.getInstance(ParticipantPersister.class);
        persistService.start();
    }

    @Test(enabled=true,groups={"unit"})
    public void getPlayerChecks() {

        Player p = new Player.Builder()
            .firstName("Rory")
            .lastName("Mcilroy")
            .build();

        // save
        p = participantPersister.savePlayer(p);
        final String playerId = p.getId();
        MatcherAssert.assertThat(playerId,Matchers.notNullValue());

        // re-read
        p = participantPersister.getPlayer(playerId);
        MatcherAssert.assertThat(p.getId(),Matchers.equalTo(playerId));
    }

    @Test(enabled=true,groups={"unit"})
    public void savePlayerChecks() {

        Player p = new Player.Builder()
            .firstName("Philip")
            .lastName("Mickelson")
            .build();

        // save
        p = participantPersister.savePlayer(p);
        MatcherAssert.assertThat(p.getId(),Matchers.notNullValue());
        MatcherAssert.assertThat(p.getLastName(),Matchers.equalTo("Mickelson"));

        // save again
        p.setFirstName("Phil");
        p = participantPersister.savePlayer(p);
        MatcherAssert.assertThat(p.getFirstName(), Matchers.equalTo("Phil"));
    }

    @Test(enabled=true,groups={"unit"})
    public void deletePlayerChecks() {

        Player p = new Player.Builder()
            .firstName("Tiger")
            .lastName("Woods")
            .build();

        // save
        p = participantPersister.savePlayer(p);
        MatcherAssert.assertThat(p.getId(),Matchers.notNullValue());

        // delete
        final String playerId = p.getId();
        participantPersister.deletePlayer(p);

        // read
        p = participantPersister.getPlayer(playerId);
        MatcherAssert.assertThat(p,Matchers.nullValue());
    }
}
```
The good news is the savePlayerChecks() passes fine. The not so good news is the getPlayerChecks() coughs up this stacktrace:
```
java.lang.NullPointerException
	at com.nextlevel.repo.ParticipantPersisterTests.getPlayerChecks(ParticipantPersisterTests.java:60)
```
And the deletePlayerChecks() gives more interesting stacktrace:
```
ru.vyarus.guice.persist.orient.repository.core.MethodExecutionException: Failed to execute repository method PlayerRepository#delete(<T>) with arguments [Player{id=#30:-2, type=PLAYER, firstName=Tiger, lastName=Woods}]
	at ru.vyarus.guice.persist.orient.repository.RepositoryMethodInterceptor.executeMethod(RepositoryMethodInterceptor.java:70)
	at ru.vyarus.guice.persist.orient.repository.RepositoryMethodInterceptor.invoke(RepositoryMethodInterceptor.java:46)
	at com.nextlevel.domain.repo.ParticipantPersister.deletePlayer(ParticipantPersister.java:53)
	at ru.vyarus.guice.persist.orient.db.transaction.internal.TransactionInterceptor$1.execute(TransactionInterceptor.java:44)
	at ru.vyarus.guice.persist.orient.db.transaction.template.TxTemplate.doInTransaction(TxTemplate.java:58)
	at ru.vyarus.guice.persist.orient.db.transaction.internal.TransactionInterceptor.invoke(TransactionInterceptor.java:41)
	at com.nextlevel.repo.ParticipantPersisterTests.deletePlayerChecks(ParticipantPersisterTests.java:96)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:84)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:714)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:901)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1231)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:127)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:111)
	at org.testng.TestRunner.privateRun(TestRunner.java:767)
	at org.testng.TestRunner.run(TestRunner.java:617)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:348)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:343)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:305)
	at org.testng.SuiteRunner.run(SuiteRunner.java:254)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1224)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1149)
	at org.testng.TestNG.run(TestNG.java:1057)
	at org.testng.remote.RemoteTestNG.run(RemoteTestNG.java:111)
	at org.testng.remote.RemoteTestNG.initAndRun(RemoteTestNG.java:204)
	at org.testng.remote.RemoteTestNG.main(RemoteTestNG.java:175)
	at org.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:125)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:134)
Caused by: ru.vyarus.guice.persist.orient.repository.delegate.DelegateMethodException: Failed to invoke delegate method ObjectVertexCrudDelegate#delete(<T>)
	at ru.vyarus.guice.persist.orient.repository.delegate.DelegateMethodExtension.execute(DelegateMethodExtension.java:68)
	at ru.vyarus.guice.persist.orient.repository.delegate.DelegateMethodExtension.execute(DelegateMethodExtension.java:24)
	at ru.vyarus.guice.persist.orient.repository.RepositoryMethodInterceptor.executeMethod(RepositoryMethodInterceptor.java:68)
	... 36 more
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at ru.vyarus.guice.persist.orient.repository.delegate.DelegateMethodExtension.execute(DelegateMethodExtension.java:66)
	... 38 more
Caused by: java.lang.NullPointerException
	at ru.vyarus.guice.persist.orient.support.repository.mixin.graph.delegate.ObjectVertexCrudDelegate.delete(ObjectVertexCrudDelegate.java:37)
	at ru.vyarus.guice.persist.orient.support.repository.mixin.graph.delegate.ObjectVertexCrudDelegate.delete(ObjectVertexCrudDelegate.java:42)
	... 43 more
```
I realize that these issues could be my misunderstanding on how the framework or OrientDB works and probably just need a little more information.

**Comment by xvik** [comment]
Could you please send me updated acme project so I could easily reproduce these issues?

**Comment by timcan** [comment]
Done, just sent it to you. Thanks.

**Comment by timcan** [comment]
Hi Vy. I've pushed the [Acme Project](https://github.com/timcan/guice-persist-orient-acme) to GitHub so that I can easily push code to you as I find other issues. I've tested Acme with your latest commit (8a7783c) and updated to use generic-resolver 2.0.0 and still having issues with the ObjectVertexCrud operations. I hope the project will also serve as Maven based tutorial with examples of usages of your framework.

**Comment by xvik** [comment]
Great! I will look better tomorrow and answer you.
I add snapshots section to readme (https://github.com/xvik/guice-persist-orient#snapshots).. maybe it would be easier than building manually.

**Comment by xvik** [comment]
Your tests will pass with the latest trunk.
Your issue was: when you create new object orient generates temporal rid (e.g. #12:-2) and when you detach object under transaction this rid stays in pojo forever. Now crud detach method tracks new object id and updates them on commit in detached pojo.

**Comment by timcan** [comment]
For anyone else having for this issue when using TestNG and the @Guice annotation is to create a method that programmatically creates the Injector and starts the persistence service:
```
 @BeforeClass(alwaysRun = true)
    public void configureInjector() {
        Injector injector = Guice.createInjector(new PersistenceModule());
        persistService = injector.getInstance(PersistService.class);
        participantPersister = injector.getInstance(ParticipantPersister.class);
        persistService.start();
    }
```
I tried using the @Guice annotation with the 3.1.0 release and OrientDB 2.0.12 still got a stacktrace on a unit test:
```
Jul 05, 2015 11:32:32 AM com.orientechnologies.common.log.OLogManager log
INFO: OrientDB auto-config DISKCACHE=10,695MB (heap=3,641MB os=16,384MB disk=137,833MB)
java.lang.AbstractMethodError: co.acme.repo.CoyoteRepository.save(Ljava/lang/Object;)Ljava/lang/Object;
	at co.acme.dao.CharacterPersister.saveCoyote(CharacterPersister.java:29)
	at ru.vyarus.guice.persist.orient.db.transaction.internal.TransactionInterceptor$1.execute(TransactionInterceptor.java:44)
	at ru.vyarus.guice.persist.orient.db.transaction.template.TxTemplate.doInTransaction(TxTemplate.java:58)
	at ru.vyarus.guice.persist.orient.db.transaction.internal.TransactionInterceptor.invoke(TransactionInterceptor.java:41)
	at co.acme.dao.CharacterPersisterTests.deleteCoyote_shouldNotExistAfterDelete(CharacterPersisterTests.java:83)
```

**Comment by xvik** [comment]
Explanation: when @ProvidedBy is used for repositories it mean JIT bindings are used for repository creation. Guice JIT binding is created in parent injector when aop interceptors are not defined. As a result, you see abstract method call error.

There are 2 solutions for this:
1. If your bean depends on other bean in child injector, then JIT binding will be created in child(!) injector. So, for example, if you add @Inject TransactionManager manager; into repository class (have to be abstract class not interface) repository will be created in child injector and aop will work.
2. Other option is to explicitly define repository (both interface/abstract class and generated implementation!). For example:
```java
final Class impl = DynamicClassGenerator.generate(MyRepository.class, Singleton.class);
bind(MyRepository.class).to(impl);
bind(impl);
```
Implementation (generated class) binding is required because otherwise it will be jit binding (and bean again will be propagated to parent injector).

Technically I could solve the problem with repositories by automatically adding some dummy dependency during class generation, but I don't like this solution. I will think about it a bit later, but can't promise I'll do it (its too hacky).

**Comment by xvik** [comment]
Fixed. Will be released in 3.2.0
